src/promptflow-tools/promptflow/tools/open_source_llm.py

```diff
@@ -0,0 +1,466 @@
+    def deco_retry(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            delay = initial_delay
+            for i in range(max_retries):
+                try:
+                    return func(*args, **kwargs)
+                except HTTPError as e:
+                    if i == max_retries - 1:
+                        raise Exception("Func execution failed after {0} retries: {1}".format(max_retries, e))
+                    delay *= exponential_base
+                    delay_from_error_message = None
+                    final_delay = delay_from_error_message if delay_from_error_message else delay
+                    time.sleep(final_delay)
+        return wrapper
+    return deco_retry
+
+
+class ModelFamily(str, Enum):
+    LLAMA = "LLaMa"
+    DOLLY = "Dolly"
+    GPT2 = "GPT-2"
+    FALCON = "Falcon"
+
+
+class API(str, Enum):
+    CHAT = "chat"
+    COMPLETION = "completion"
+
+
+class ContentFormatterBase:
+    """Transform request and response of AzureML endpoint to match with
+    required schema.
+    """
+
+    content_type: Optional[str] = "application/json"
+    """The MIME type of the input data passed to the endpoint"""
+
+    accepts: Optional[str] = "application/json"
+    """The MIME type of the response data returned from the endpoint"""
+
+    @staticmethod
+    def escape_special_characters(prompt: str) -> str:
+        """Escapes any special characters in `prompt`"""
+        escape_map = {
+            "\\": "\\\\",
+            '"': '\\"',
+            "\b": "\\b",
+            "\f": "\\f",
+            "\n": "\\n",
+            "\r": "\\r",
+            "\t": "\\t",
+        }
+
+        # Replace each occurrence of the specified characters with their escaped versions
+        for escape_sequence, escaped_sequence in escape_map.items():
+            prompt = prompt.replace(escape_sequence, escaped_sequence)
+
+        return prompt
+
+    @abstractmethod
+    def format_request_payload(self, prompt: str, model_kwargs: Dict) -> bytes:
+        """Formats the request body according to the input schema of
+        the model. Returns bytes or seekable file like object in the
+        format specified in the content_type request header.
+        """
+
+    @abstractmethod
+    def format_response_payload(self, output: bytes) -> str:
+        """Formats the response body according to the output
+        schema of the model. Returns the data type that is
+        received from the response.
+        """
+
+
+class GPT2ContentFormatter(ContentFormatterBase):
+    """Content handler for LLMs from the OSS catalog."""
+
+    def format_request_payload(self, prompt: str, model_kwargs: Dict) -> bytes:
+        input_str = json.dumps(
+            {
+                "inputs": {"input_string": [ContentFormatterBase.escape_special_characters(prompt)]},
+                "parameters": model_kwargs,
+            }
+        )
+        return str.encode(input_str)
```

Comment:
Do we need encoding here? Will this encode fail?

Reply:
Yes - it seems to be a limitation on the Foundation model score.py files. I can make it more generic though.